mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -460,21 +486,52 @@ private[ml] object SummaryBuilderImpl extends Logging {
       val realMean = Array.ofDim[Double](n)
       var i = 0
       while (i < n) {
-        realMean(i) = currMean(i) * (weightSum(i) / totalWeightSum)
+        realMean(i) = currMean(i) * (currWeightSum(i) / totalWeightSum)
         i += 1
       }
       Vectors.dense(realMean)
     }
 
+    /**
+     * Sum of each dimension.
+     */
+    def sum: Vector = {
+      require(requestedMetrics.contains(Sum))
+      require(totalWeightSum > 0, s"Nothing has been added to this summarizer.")
+
+      val realSum = Array.ofDim[Double](n)
+      var i = 0
+      while (i < n) {
+        realSum(i) = currMean(i) * currWeightSum(i)
+        i += 1
+      }
+      Vectors.dense(realSum)
+    }
+
     /**
      * Unbiased estimate of sample variance of each dimension.
      */
     def variance: Vector = {
       require(requestedMetrics.contains(Variance))
       require(totalWeightSum > 0, s"Nothing has been added to this summarizer.")
 
-      val realVariance = Array.ofDim[Double](n)
+      val realVariance = computeVariance
+      Vectors.dense(realVariance)
+    }
+
+    /**
+     * Unbiased estimate of standard deviation of each dimension.
+     */
+    def std: Vector = {
+      require(requestedMetrics.contains(Std))
+      require(totalWeightSum > 0, s"Nothing has been added to this summarizer.")
 
+      val realVariance = computeVariance
+      Vectors.dense(realVariance.map(math.sqrt))
+    }
+
+    private def computeVariance: Array[Double] = {
+      val realVariance = Array.ofDim[Double](n)
       val denominator = totalWeightSum - (weightSquareSum / totalWeightSum)
```

Comment:
Whatever this is, I'm not sure why it would be negative, unless the idea is simply to account for floating-point imprecision. But if it is the equivalent of diving by "n-1" then this seems to be incorrectly computing 'sample variance'. I'm not quite sure how sample variance is supposed to work out in the weighted case, but, if that's what it's doing it's not correct anyway, I think.

It would then be valid to require totalWeightSum > 0, but equally, would be valid and consistent with pandas to return NaN I think. But we can keep the current behavior.

Reply:
The idea behind this `def computeVariance: Array[Double]` method is only to reuse the code.
Since according current design, we will check `requestedMetrics.contains(...)` before computation, so in `def std` we can not safely obtain a variance vector by call `def variance`

Reply:
Ooh, you mean `We prevent variance from negative value caused by numerical error.`
yes, I guess it is reasonable to del this check, it should not have a negative denominator.

Reply:
Maybe this is for numeric-stablity?

Reply:
Yeah probably. Well, per above, I think it's doing what we want anyway. We can leave the logic as-is.